frame/curve-amm/src/lib.rs

```diff
@@ -1011,53 +656,139 @@ pub mod pallet {
 			None
 		}
 
-		fn transfer_liquidity_into_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source: &T::AccountId,
-			destination_asset_index: usize,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[destination_asset_index],
-				source,
-				pool_account_id,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[destination_asset_index] = pool.total_balances
-				[destination_asset_index]
-				.checked_add(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
-		fn transfer_liquidity_from_pool(
-			pool_account_id: &T::AccountId,
-			pool: &mut PoolInfo<T::AccountId, T::AssetId, T::Balance>,
-			source_asset_index: usize,
-			destination: &T::AccountId,
-			amount: T::Balance,
-		) -> DispatchResult {
-			T::LpToken::transfer(
-				pool.assets[source_asset_index],
-				pool_account_id,
-				destination,
-				amount,
-				true,
-			)?;
-
-			pool.total_balances[source_asset_index] = pool.total_balances[source_asset_index]
-				.checked_sub(&amount)
-				.ok_or(Error::<T>::InconsistentStorage)?;
-
-			Ok(())
-		}
-
 		pub fn account_id(pool_id: &PoolId) -> T::AccountId {
 			PALLET_ID.into_sub_account(pool_id)
 		}
+
+        fn limit_preserving_invariant( 
+			pool_id: PoolId,
+            index : usize,
+            min_balance : T::Balance,
+            max_balance : T::Balance,
+        ) -> Result<T::Balance, DispatchError> {
+            let value = Pools::<T>::try_mutate(pool_id, |pool| -> Result<T::Balance, DispatchError> {
+                let pool = pool.as_mut().ok_or(Error::<T>::PoolNotFound)?;
+					let n_coins = pool.assets_vault_ids.len();
+
+                let pool_account_id = Self::account_id(&pool_id);
+                let pool_asset_ids: Vec<T::AssetId> = pool
+                    .assets_vault_ids
+                    .iter()
+                    .map(|vault_id| {
+                        Ok(T::Vault::asset_id(vault_id).ok().ok_or(Error::VaultError)?.into())
+                    })
+                .collect::<Result<Vec<T::AssetId>, Error<T>>>()?;
+                let mut balances = Vec::new();
+                for asset_id in &pool_asset_ids {
+                    let bal = T::LpToken::balance(*asset_id, &pool_account_id);
+                    balances.push(bal);
+                }
+                let mut xp_f: Vec<FixedU128> = balances
```

Comment:
not sure i get what xp_f means? is it from paper?

Reply:
ah, it is fixed? do we need non fixed? could we remove all _f things?

Reply:
all computations are done in FixedU128 , but in runtime we can;t use it directly so in runtime we use u128 (T::Balance). This is similar to what we have done in lending.